### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by neremic**
Hi,

I rewrote most of the components in a better fashion. 
This pull request is just about those changes. Tests are failing and need to be adjusted.

So please just give feedback about the components - meaning everything in /src/components/

Thanks in advance and cheers

### CODE REVIEW

File: src/components/functional/Chart.jsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
+"use strict";
```
Comment (prayerslayer) [inline src/components/functional/Chart.jsx]: You don't need this as ES6 modules [are always strict mode](https://tc39.github.io/ecma262/#sec-strict-mode-code).

File: src/components/functional/Chart.jsx [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+const SIDE_STYLE = {display: "flex", flexDirection: "column", width: 150};
+const MIDDLE_STYLE = {display: "flex", flexDirection: "column", flex: "auto"};
+
+class Chart extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {selectedElements: new Set()};
+
+        this.WrappedAreaChart = Hideable(AreaChart);
```
Comment (prayerslayer) [inline src/components/functional/Chart.jsx]: Maybe `HideableAreaChart` is a better name. `Wrapped` doesn't tell me anything :)

File: src/components/functional/Chart.jsx [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+    getEventDataByEventNames() {
+        return this.props.eventData.filter(d => this.state.selectedElements.has(d.name));
+    }
+
+    render() {
+        const { version } = this.props;
+        const filteredEventData = this.getEventDataByEventNames();
+
+        let childrenWithProps = [];
+        const childWithProps = React.Children.map(this.props.children,
```
Comment (prayerslayer) [inline src/components/functional/Chart.jsx]: You use already `map`, no need to `push` in a separate array?

File: src/components/functional/Chart.jsx [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                        <TitleWithButton
+                            title   = {version.text}
+                            onClick = {this.props.onDelete}
+                        />
+                    </div>
+                    <div>
+                        <SelectableLabelList
+                            elements         = {this.props.eventData}
+                            selectedElements = {this.state.selectedElements}
+                            select           = {this.elementSelected}
```
Comment (prayerslayer) [inline src/components/functional/Chart.jsx]: I like the `on...` naming convention for callbacks.

File: src/components/functional/Chart.jsx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+                <div style={SIDE_STYLE}>
+                    {childrenWithProps}
+                </div>
+            </div>
+        )
+    }
+}
+
+const dataPending = function(props) {
+    return props.dataSet == undefined || props.dataSet.values.length == 0;
```
Comment (prayerslayer) [inline src/components/functional/Chart.jsx]: What happens when there just is no data?

File: src/components/functional/GlobalBrush.jsx [hunk 6]
```diff
@@ -50,7 +52,7 @@ class GlobalBrush extends React.Component {
                         height = {this.props.height}
                         margin = {BRUSH_MARGIN}
                         xScale = {d3.time.scale().domain([startDate, endDate]).range([0, this.props.containerWidth - BRUSH_HORIZONTAL_MARGIN])}
-                        extent = {[startDate, endDate]}
+                        extent = {[this.state.extent[0], this.state.extent[1]]}
```
Comment (prayerslayer) [inline src/components/functional/GlobalBrush.jsx]: Not sure I get this `extent` variable. Since it's only ever `[startDate, endDate]` you can use those directly **or** at least shorten this to `extent={this.state.extent}`.

File: src/components/functional/VersionSelector.jsx [hunk 7]
```diff
@@ -31,7 +33,7 @@ class VersionSelector extends React.Component {
     componentWillReceiveProps(nextProps) {
         let rVersion = nextProps.removeVersion;
         if (rVersion !== undefined && rVersion !== this.state.lastRemovedVersion) {
-            let newValues = this.state.values.filter((v) => {
+            const newValues = this.state.values.filter((v) => {
```
Comment (prayerslayer) [inline src/components/functional/VersionSelector.jsx]: Since a version is presumably unique: Why can't you just make sure that `versions` does not contain `removeVersion` or overwrite `values` and letting the parent component handle the removal?

File: src/components/hoc/LoadingIf.jsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        super(props);
+        this.state = {
+            dots: '',
+            loading: loadingCondition(props)
+        }
+    }
+
+    componentDidMount() {
+        this.interval = setInterval(() => {
+            let currentDots = this.state.dots + '.';
```
Comment (prayerslayer) [inline src/components/hoc/LoadingIf.jsx]: No offense, but since you're already using bootstrap for your prototype (which we should remove anyways), wasn't there any spinning icon support? Like this: http://fontawesome.io/examples/#animated

Comment (prayerslayer) [inline src/components/hoc/LoadingIf.jsx]: So maybe you could add support for a custom loading indicator and leave this as a default or smth.

File: src/components/application/life_cycle_charts/life_cycle_charts.jsx [hunk 9]
```diff
@@ -119,65 +147,70 @@ class LifeCycleCharts extends React.Component {
...
-                            selectedVersions = {this.state.selectedVerions}
-                            onDelete = {this.handleVersionRemoved}
-                            dataSets = {this.props.versionsHistory}
-                            viewPortDateRange = {this.state.zoomedViewPortDateRange}
-                            applicationId = {this.state.applicationId}
-                            height = {CHART_HEIGHT}
-                        />
-                    </div>
-                </div>
+                <Charts
```
Comment (prayerslayer) [inline src/components/application/life_cycle_charts/life_cycle_charts.jsx]: What is the `PropsExposer` for, why can't you use `react-dimensions` directly? They have some docs on how to use it with flexbox.

File: src/components/pure/ScatterChartRD3.jsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
+"use strict";
+
+import React from 'react';
+import { ScatterChart } from 'rd3';
+
+const FORMATTER_ONLY_INTEGERS = function(number) {
+    return number == Math.floor(number) ? number : null;
```
Comment (prayerslayer) [inline src/components/pure/ScatterChartRD3.jsx]: `Number.isInteger`?

File: src/components/pure/ScatterChartRD3.jsx [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            :
+            <div style = {{width : width, height : height}}>
+                <h1>
+                    Nothing to show
+                </h1>
+            </div>
+    )
+};
+
+ScatterChartRD3.PorpTypes = {
```
Comment (prayerslayer) [inline src/components/pure/ScatterChartRD3.jsx]: typo

File: src/components/util/colorUtils.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
+"use strict";
+
+const BASE_16 = 16;
+
+const calculateColors = function(elements) {
```
Comment (prayerslayer) [inline src/components/util/colorUtils.js]: Please read up on D3's `color scales`. You don't have to do this :D

Example: https://jsfiddle.net/vpf5t33z/

**Comment by prayerslayer** [comment]
Regarding the API:

## /lifecycle/events/apps/{application_id}/versions

* events could have a "source", e.g. aws account + service, to account for GCP in the future
* events should also have a "resource", e.g. instance id for RunInstances. makes it immediately more useful when used together with aws web ui
* "change_date" probably a misnomer, also description misleading
* what's the "sort" parameter doing?

## /lifecycle/life_change_events/apps/{application_id}/versions

* `_` in path discouraged by api guild, use `-` (fine for path parameters)
* as this yields the amount of servers, we could as well name it `instances`? not sure how this would fit with gcp kubernetes, would this be pods?
* we should find a way to include ids here, too

**Comment by harti2006** [comment]
I would not make any assumptions on the format of application or version ids:
https://github.com/neremic/zalimat/blob/92beee6648f3f46536e903bcbb1255ef46a9e94f/api.yaml#L346

**Comment by harti2006** [comment]
Some more thoughts:
* Can we remove the "/lifecycle" prefix?
* can we omit the "sort" param (events are always ordered by "event_date")?!
* should we better separate fetching version and the data for the versions?
